app/objects/c_operation.py

```diff
@@ -92,13 +90,14 @@ def __init__(self, name, agents, adversary, id=None, jitter='2/8', source=None,
         self.planner = planner
         self.state = state
         self.autonomous = autonomous
-        self.phases_enabled = phases_enabled
-        self.phase = 0
+        self.atomic_enabled = atomic_enabled
+        self.last_ran = None
         self.obfuscator = obfuscator
         self.auto_close = auto_close
         self.visibility = visibility
         self.chain, self.rules = [], []
         self.access = access if access else self.Access.APP
+        self.generate_expired = 0
```

Comment:
Can you explain the reasoning for this a bit more?  I see it's used as a bounds check for ending an operation.

Reply:
That's exactly what it is. In the case of a self-closing operation, we don't have a reliable way to determine that we've completed everything in the queue now that we've removed phases (we can't check that we completed the "last" phase). This is a counter of the number of times the planner has failed to generate links, which, when combined with the "global" step list for the adversary, implies that the planner can't find anything else to do. It'll time out after this occurs 3 times, closing the operation.

Reply:
this is the only part that I'm not totally onboard with. 

there are two things at play here, generate_expired + last_run (ability). they're both created and used only to determine when to close an operation (or "phase"). I haven't put cycles into how to consolidate these and make it more simplistic - but if we can try something else, the rest of this PR looks pretty good.